### PR TITLE
Feat/torch

### PR DESCRIPTION
**Pull request by dilyararimovna**
- `torch_trainer` which is needed to `model.train()` and `model.eval()` settings. By default, torch models are set to `train` mode, so for `build_model` correct usage we still need to do `eval` in `init` of each torch model. We can add try `model.eval` everywhere in `build_model`, `interact` mode, etc but it looks very kostyl'no.
- base class `TorchModel` which saves, loads, initializes model on pytorch.
- `torch_classification_model` as an example of Neural Network implementation for classification task.
- BERT preprocessors on torch are almost all implemented
- BERT-based classification model on torch (with config example)
- BERT-based ranker model on torch (with config example)
- BERT-based ner model on torch (with config example)
- BERT-based SQuAD model on torch (with config example)
- BERT-based summarization model on torch (with config example)
- files with example models from configs are not all ready (will be uploaded to share later)

### CODE REVIEW

File: deeppavlov/requirements/torch.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+torch==1.5.1
+torchtext==0.6.0
+torchsummary==1.5.1
+transformers==3.0.2
```
Comment (kudep) [inline deeppavlov/requirements/torch.txt:4]: All are in one files, it's norm? @yurakuratov @dilyararimovna @IgnatovFedor

File: deeppavlov/core/models/torch_model.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            self.load_path = fname
+
+        model_func = getattr(self, self.opt.get("model_name"), None)
+
+        if self.load_path:
+            log.info(f"Load path {self.load_path} is given.")
+            if isinstance(self.load_path, Path) and not self.load_path.parent.is_dir():
+                raise ConfigError("Provided load path is incorrect!")
+
+            weights_path = Path("{}.pth.tar".format(str(self.load_path.resolve())))
```
[the remark this answers is not in the document]
Comment (dilyararimovna) [reply deeppavlov/core/models/torch_model.py:116]: ты наверно хотел написать что-то такое:
```
weights_path = Path(self.load_path.resolve())
weights_path = weights_path.with_suffix(f".pth.tar")
```
поправила у себя

File: deeppavlov/core/trainers/torch_trainer.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+class TorchTrainer(NNTrainer):
+
+    def test(self, data: Iterable[Tuple[Collection[Any], Collection[Any]]],
+             metrics: Optional[Collection[Metric]] = None, *,
+             start_time: Optional[float] = None, show_examples: Optional[bool] = None) -> dict:
+        self._chainer.get_main_component().model.eval()
+
+        report = super(TorchTrainer, self).test(data=data, metrics=metrics, start_time=start_time,
+                                                show_examples=show_examples)
+        self._chainer.get_main_component().model.train()
```
[the remark this answers is not in the document]
Comment (dilyararimovna) [reply deeppavlov/core/trainers/torch_trainer.py:36]: мы по дефолту хотим, чтобы модель была в eval (а сами pytorch models инитятся в режиме train, поэтому в инит torch_model есть перевод в eval mode).
Но тут начинается трена, и чтобы в каждом батче не ставить ее в `train` режим, я тут в начале трены ставлю в train, а в конце обратно в eval. Причем в начале test функции (валидация и эвалюация в конце) я ставлю в eval режим, а в конце ее в train обратно.

File: deeppavlov/requirements/torch.txt (hunk 1 above)
Comment (dilyararimovna) [reply deeppavlov/requirements/torch.txt]: заиспользовала уже имеющиеся файлы с зависимостями.

File: deeppavlov/core/models/torch_model.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+class TorchModel(NNModel):
+    """Class implements torch model's main methods.
+
+    Args:
+        device: `cpu` or `cuda` device to use
+        learning_rate_drop_patience: how many validations with no improvements to wait
+        learning_rate_drop_div: the divider of the learning rate after `learning_rate_drop_patience` unsuccessful
+            validations
+        load_before_drop: whether to load best model before dropping learning rate or not
```
[the remark this answers is not in the document]
Comment (dilyararimovna) [reply deeppavlov/core/models/torch_model.py:42]: described

File: deeppavlov/core/trainers/torch_trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kudep) [reply deeppavlov/core/trainers/torch_trainer.py:36]: Но тут ты все равно ставить `train`  режим https://github.com/deepmipt/DeepPavlov/blob/174cc643a6a347fb60099bf7c1f6553feb32c634/deeppavlov/core/trainers/torch_trainer.py#L40

File: deeppavlov/core/models/torch_model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kudep) [reply deeppavlov/core/models/torch_model.py]:
```python
weights_path = self.load_path.resolve()
weights_path = weights_path.with_suffix(f"{weights_path.suffix}.pth.tar")
```
иначе суффикс, который был до этого потеряем

File: deeppavlov/core/trainers/torch_trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kudep) [reply deeppavlov/core/trainers/torch_trainer.py:36]: Для каждого батча

[the remark this answers is not in the document]
Comment (dilyararimovna) [reply deeppavlov/core/trainers/torch_trainer.py:36]: это не по батчам функция вызывается

[the remark this answers is not in the document]
Comment (dilyararimovna) [reply deeppavlov/core/trainers/torch_trainer.py:36]:
```
def train_on_batches(self, iterator: DataLearningIterator) -> None:
        self._chainer.get_main_component().model.train()
        super(TorchTrainer, self).train_on_batches(iterator=iterator)
        self._chainer.get_main_component().model.eval()
```
она внутри себя идет по итератору по батчам. А сама функция `train_on_batches` вызывается один раз.

File: deeppavlov/core/models/torch_model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dilyararimovna) [reply deeppavlov/core/models/torch_model.py]: а зачем он нам?)

[the remark this answers is not in the document]
Comment (kudep) [reply deeppavlov/core/models/torch_model.py]: На тот случай если будет

[the remark this answers is not in the document]
Comment (dilyararimovna) [reply deeppavlov/core/models/torch_model.py]: я просто считаю, что там не должно быть никаких еще суффиксов

File: deeppavlov/models/torch_bert/torch_bert_classifier.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        if self.lr_scheduler_name is not None:
+            self.lr_scheduler = getattr(torch.optim.lr_scheduler, self.lr_scheduler_name)(
+                self.optimizer, **self.lr_scheduler_parameters)
+
+        if self.load_path:
+            log.info(f"Load path {self.load_path} is given.")
+            if isinstance(self.load_path, Path) and not self.load_path.parent.is_dir():
+                raise ConfigError("Provided load path is incorrect!")
+
+            weights_path = Path("{}.pth.tar".format(str(self.load_path.resolve())))
```
Comment (kudep) [inline deeppavlov/models/torch_bert/torch_bert_classifier.py:187]: same issue

File: deeppavlov/models/torch_bert/torch_bert_ranker.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+        Args:
+            features_li: list of elements where each element contains the batch of features
+             for contexts with particular response candidates
+
+        Returns:
+            predicted scores for contexts over response candidates
+        """
+        if len(features_li) == 1 and len(features_li[0]) == 1:
+            msg = "It is not intended to use the {} in the interact mode.".format(self.__class__)
```
Comment (kudep) [inline deeppavlov/models/torch_bert/torch_bert_ranker.py:103]: I propose to replace format by f-strings

File: deeppavlov/dataset_readers/squad_dataset_reader.py [hunk 7]
```diff
@@ -75,8 +75,7 @@ def read(self, dir_path: str, dataset: Optional[str] = 'SQuAD', url: Optional[st
 
         dir_path = Path(dir_path)
         required_files = ['{}-v1.1.json'.format(dt) for dt in ['train', 'dev']]
```
Comment (kudep) [inline deeppavlov/dataset_readers/squad_dataset_reader.py:77]: f-strings

File: deeppavlov/models/torch_bert/torch_bert_classifier.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+            logits = logits.detach().cpu().numpy()
+            pred = np.argmax(logits, axis=1)
+        return pred
+
+    @overrides
+    def load(self, fname=None):
+        if fname is not None:
+            self.load_path = fname
+
+        if self.pretrained_bert and not os.path.isfile(self.pretrained_bert):
```
Comment (kudep) [inline deeppavlov/models/torch_bert/torch_bert_classifier.py:159]: in case likes that we can use `Path` from `pathlib` and remove `import os`

**Comment by kudep** [comment]
There are good at all, I find a little bit moments, but they don't matter in general.

File: deeppavlov/models/torch_bert/torch_bert_classifier.py (hunk 8 above)
Comment (kudep) [reply deeppavlov/models/torch_bert/torch_bert_classifier.py:159]: if you want you can fix it. For it's not important.

File: deeppavlov/models/torch_bert/torch_bert_sequence_tagger.py [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+    """
+
+    def __init__(self,
+                 n_tags: int,
+                 pretrained_bert: str,
+                 bert_config_file: Optional[str] = None,
+                 return_probas: bool = False,
+                 attention_probs_keep_prob: Optional[float] = None,
+                 hidden_keep_prob: Optional[float] = None,
+                 optimizer: str = None,
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_bert_sequence_tagger.py:226]: Defalt value for base class is `AdamW`. Won't default value `None` cause Error in base class?

File: deeppavlov/core/models/torch_model.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+        self.optimizer_name = optimizer
+        self.optimizer_parameters = optimizer_parameters
+        self.lr_scheduler_name = lr_scheduler
+        self.lr_scheduler_parameters = lr_scheduler_parameters
+
+        self.learning_rate_drop_patience = learning_rate_drop_patience
+        self.learning_rate_drop_div = learning_rate_drop_div
+        self.load_before_drop = load_before_drop
+        self.min_learning_rate = min_learning_rate if min_learning_rate is not None else 0.
```
Comment (IgnatovFedor) [inline deeppavlov/core/models/torch_model.py:90]:
```suggestion
        self.min_learning_rate = min_learning_rate
```

File: deeppavlov/models/torch_bert/torch_bert_sequence_tagger.py [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+                 optimizer_parameters: dict = {"lr": 1e-3, "weight_decay": 1e-6},
+                 learning_rate_drop_patience: int = 20,
+                 learning_rate_drop_div: float = 2.0,
+                 load_before_drop: bool = True,
+                 clip_norm: Optional[float] = None,
+                 min_learning_rate: Optional[float] = 1e-07,
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_bert_sequence_tagger.py:232]:
```suggestion
                 min_learning_rate: float = 1e-07,
```

File: deeppavlov/models/torch_bert/torch_bert_squad.py [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+import torch
+from transformers import BertForQuestionAnswering, BertConfig, BertTokenizer
+from transformers.data.processors.utils import InputFeatures
+
+from deeppavlov import build_model
+from deeppavlov.core.common.errors import ConfigError
+from deeppavlov.core.commands.utils import expand_path
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.models.estimator import Component
+from deeppavlov.core.models.torch_model import TorchModel
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_bert_squad.py:33]:
```suggestion
import json
import math
import os
from logging import getLogger
from pathlib import Path
from typing import List, Tuple, Optional, Dict

import numpy as np
import torch
from overrides import overrides
from transformers import BertForQuestionAnswering, BertConfig, BertTokenizer
from transformers.data.processors.utils import InputFeatures

from deeppavlov import build_model
from deeppavlov.core.commands.utils import expand_path
from deeppavlov.core.common.errors import ConfigError
from deeppavlov.core.common.registry import register
from deeppavlov.core.models.estimator import Component
from deeppavlov.core.models.torch_model import TorchModel
```

File: deeppavlov/models/torch_bert/torch_bert_classifier.py (hunk 8 above)
Comment (dilyararimovna) [reply deeppavlov/models/torch_bert/torch_bert_classifier.py:159]: fixed

File: deeppavlov/core/models/torch_model.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+# limitations under the License.
+
+from abc import abstractmethod
+from typing import Optional
+from logging import getLogger
+from pathlib import Path
+from copy import deepcopy
+from overrides import overrides
+
+import torch
```
[the remark this answers is not in the document]
Comment (dilyararimovna) [reply deeppavlov/core/models/torch_model.py:22]: no commit here but fixed (overrides before torch should be)

File: deeppavlov/models/torch_bert/torch_bert_sequence_tagger.py (hunk 9 above)
Comment (dilyararimovna) [reply deeppavlov/models/torch_bert/torch_bert_sequence_tagger.py:226]: will.fixed
